File: Sort/shell_sort.go
package main

import (
	"fmt"
	"sort"
)

func ShellSort(data sort.Interface) {
	n := data.Len()
	gap := n / 2
	for gap > 0 {
		for i := gap; i < n; i++ {
			for j := i; j-gap >= 0 && data.Less(j, j-gap); {
				data.Swap(j, j-gap)
				j -= gap
			}
		}
		gap = gap / 2
	}
}

func main() {
	unorder := sort.IntSlice{9, 8, 7, 6, 5, 4, 3, 2, 1, 0}
	fmt.Printf("before order: %v\n", unorder)

	ShellSort(unorder)

	fmt.Printf("after order: %v\n", unorder)
}

File: Sort/shell_sort.cpp
#include <array>
#include <iostream>

using namespace std;

template<class RandomIt>
void shell_sort(RandomIt first, RandomIt last) {
    int length = last - first;
    if (length <= 0) {
        return;
    }
    int gap = length / 2;
    while (gap > 0) {
        for (RandomIt i = first+gap; i != last; ++i) {
            auto val = *i;
            RandomIt j = i-gap;
            while (j >= first && *j > val) {
                *(j+gap) = *j;
                j -= gap;
            }
            *(j+gap) = val;
        }
        gap /= 2;
    }
    return;
}

template<class It>
void printByIt(It begin, It end) {
    for (It curr = begin; curr != end; ++curr) {
        cout << *curr << " ";
    }
    cout << endl;
}

int main() {
    array<int, 10> unorder = {9, 8, 7, 6, 5, 4, 3, 2, 1};
    cout << "before order: ";
    printByIt(unorder.begin(), unorder.end());

    shell_sort(unorder.begin(), unorder.end());

    cout << "after order: ";
    printByIt(unorder.begin(), unorder.end());
}
